Stock.Api/Controllers/ProviderController.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Stock.Api.DTOs;
+using Stock.Api.Extensions;
+using Stock.AppService.Services;
+using Stock.Model.Entities;
+
+namespace Stock.Api.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/provider")]
+    [ApiController]
+    public class ProviderController : ControllerBase
+    {
+        private ProviderService service;
+        private readonly IMapper mapper;
+        public ProviderController(ProviderService service, IMapper mapper)
+        {
+            this.service = service;
+            this.mapper = mapper;
+        }
+        /// <summary>
+        /// Permite recuperar todas las instancias
+        /// </summary>
+        [HttpGet]
+        public ActionResult<IEnumerable<ProviderDTO>> Get()
+        {
+            try
+            {
+                var result = this.service.GetAll();
+                return this.mapper.Map<IEnumerable<ProviderDTO>>(result).ToList();
+            }
+            catch(Exception)
+            {
+                return StatusCode(500);
+            }
+        }
+        /// <summary>
+        /// Permite recuperar una instancia mediante un identificador
+        /// </summary>
+        /// <param name="id">Identificador de la instancia a recuperar</param>
+        /// <returns>Una instancia</returns>
+        [HttpGet("{id}")]
+        public ActionResult<ProviderDTO> Get(string id)
+        {
+            try
+            {
+                var result = this.service.Get(id);
+                return this.mapper.Map<ProviderDTO>(result);
+            }
+            catch (Exception)
+            {
+                return StatusCode(500);
+            }
+        }
+        /// <summary>
+        /// Permite crear un provider
+        /// </summary>
+        /// <param name ="value">Una Instancia</param> 
+        [HttpPost]
+        public ActionResult Post([FromBody] ProviderDTO value){
+            TryValidateModel(value);
+            try{
+                var provider = this.mapper.Map<Provider>(value);
+                this.service.Create(provider);
+                value.Id = provider.Id;
+                return Ok(new {Success = true, Message = "", data = value});
+            }
+            catch{
+                return Ok(new {Success = false, Message = "The name is already in used"});
+            }
```

Comment:
Que pasaría si la creación en el servicio arroja otra excepcion que no sea porque el nombre ya esta en uso? Que mensaje le estaríamos devolviendo al cliente? Crees que podrías mejorarlo?

Reply:
Si el servicio arroja otra excepción no nos informaría realmente lo que sucedió, Siempre se le enviaría el mensaje "The name is already in used" sin importar el error. Lo mejoraria agregándole un argumento al catch del tipo Exception y devolveria el mensage que me da la excepcion. 
Solución:
catch (Exception e){ 
                return Ok(new {Success = false, Message = e.Message});
            }

Reply:
Con eso que proponés el cliente ya sabría que el error no es el nombre sino otra situación ajena a él lo cual es lo mas importante, así que con eso serviría.
También ten en cuenta que puedes generar distintos catch para un mismo try diferenciando por tipo de excepcion. De esta manera, otra alternativa sería dentro del servicio, al fallar la validación en vez de devoler "throw new System.Exception("The name is already in use");" podemos devolver un  System.ArgumentException. Luego en el provider agregamos
            catch (System.ArgumentException e)
            {
               return Ok(new {Success = false, Message = e.Message});
            }
            catch
            {
                return new StatusCodeResult(500);
            }

Eso va a hacer que si encuentra una excepcion de tipo ArgumentException devuelve el ok indicando que la request se proceso correctamente pero que fallo  la validación de nombre, si es otro error va a devolver un "internal server error". Sólo tienes que tener en cuenta cuando haces esto que primero tienes que definir los catch para las excepciones más específicas porque si dejas la respuesta genérica primero nunca va a entrar por las otras.
El tema de mandar el mensaje cuando es una excepción interna es que en general se evita exponer al cliente (aplicación cliente) a detalles de errores internos, entonces simplemente le mandamos un 500 y ya sabe que la falla es de la api.

Espero no habértela complicado mucho y que se haya entendido, la solución que planteaste era suficiente igual.

Reply:
Me re sirve la explicación. Porque no tengo mucha experiencia con trabajo de Api, todo lo que sea para corregir me ayuda un montón para seguir sumando en aprendizaje.